File: apps/sourcerer-ui/src-tauri/src/commands/folders.rs
//! Folders panel — Indexes → Folders (PRD §8.13).

use crate::daemon;

#[tauri::command]
pub fn folders_list() -> Result<Vec<sourcerer_rpc::WatchedFolder>, String> {
    daemon_call("folders.list", serde_json::Value::Null)
}

#[tauri::command]
pub fn folders_add(folder: sourcerer_rpc::WatchedFolder) -> Result<(), String> {
    daemon_call_void(
        "folders.add",
        serde_json::to_value(folder).map_err(|e| e.to_string())?,
    )
}

#[tauri::command]
pub fn folders_remove(id: String) -> Result<(), String> {
    daemon_call_void("folders.remove", serde_json::json!({ "id": id }))
}

#[tauri::command]
pub fn folders_update(folder: sourcerer_rpc::WatchedFolder) -> Result<(), String> {
    daemon_call_void(
        "folders.update",
        serde_json::to_value(folder).map_err(|e| e.to_string())?,
    )
}

#[tauri::command]
pub fn folders_rescan(id: String) -> Result<(), String> {
    daemon_call_void("folders.rescan", serde_json::json!({ "id": id }))
}

#[tauri::command]
pub fn folders_rescan_all() -> Result<(), String> {
    daemon_call_void("folders.rescan_all", serde_json::Value::Null)
}

fn daemon_call<T>(method: &'static str, params: serde_json::Value) -> Result<T, String>
where
    T: for<'de> serde::Deserialize<'de> + Send + 'static,
{
    let daemon = daemon::get().ok_or_else(|| "daemon not initialized".to_string())?;
    daemon.call(method, params).map_err(|e| e.to_string())
}

fn daemon_call_void(method: &'static str, params: serde_json::Value) -> Result<(), String> {
    let daemon = daemon::get().ok_or_else(|| "daemon not initialized".to_string())?;
    daemon.call_void(method, params).map_err(|e| e.to_string())
}

File: apps/freally-ui/src/lib/ipc/folders.ts
import { invoke } from "@tauri-apps/api/core";
import type { WatchedFolder } from "./types";

export async function list(): Promise<WatchedFolder[]> {
  return invoke<WatchedFolder[]>("folders_list");
}

export async function add(folder: WatchedFolder): Promise<void> {
  return invoke<void>("folders_add", { folder });
}

export async function remove(id: string): Promise<void> {
  return invoke<void>("folders_remove", { id });
}

export async function update(folder: WatchedFolder): Promise<void> {
  return invoke<void>("folders_update", { folder });
}

export async function rescan(id: string): Promise<void> {
  return invoke<void>("folders_rescan", { id });
}

export async function rescanAll(): Promise<void> {
  return invoke<void>("folders_rescan_all");
}
